File: scripts/__pbot/PBotCharacterAPI.py
from typing import Optional, List

from .PBotItem import PBotItem

class PBotCharacterAPI(object):
    def __init__(self, session):
        self.__session = session

    ## Returns the player stamina
    # @return Player stamina between 0-100
    def get_stamina(self) -> float:
        return float(self.__session.PBotCharacterAPI().getStamina())

    ## Returns the player energy
    # @return Player energy between 0-100
    def get_energy(self) -> float:
        return float(self.__session.PBotCharacterAPI().getEnergy())

    ## Returns the player hp
    # @return Player hp between 0-100
    def get_hp(self) -> float:
        return float(self.__session.PBotCharacterAPI().getHp())

    ## Send act message to server
    # Act can be used for example to choose a cursor
    # Some acts:
    # dig, mine, carry, destroy, fish, inspect, repair, crime, swim, tracking, aggro, shoot
    # @param act list of acts
    def do_act(self, act: List[str]):
        self.__session.PBotCharacterAPI().doAct(act)

    ## Resname of the cursor currently selected
    # May be none if cursor couldn't be found
    # @return resname of the cursor
    def get_curs_name(self) -> Optional[str]:
        return self.__session.PBotCharacterAPI().getCursName()

    ## Cancels the current act by right clicking
    def cancel_act(self):
        self.__session.PBotCharacterAPI().cancelAct()

    ## Set player speed setting
    # @param speed 1 = crawl, 2 = walk, 3 = run, 4 = sprint
    def set_speed(self, speed: int):
        self.__session.PBotCharacterAPI().setSpeed(speed)

    ## Get current speed setting of player
    # @return 1 = crawl, 2 = walk, 3 = run, 4 = sprint
    def get_speed(self) -> int:
        return self.__session.PBotCharacterAPI().getSpeed()

    ## Get maximum speed setting that the player can be set to
    # @return 1 = crawl, 2 = walk, 3 = run, 4 = sprint
    def get_max_speed(self) -> int:
        return self.__session.PBotCharacterAPI().getMaxSpeed()

    ## Send message to the given chat
    # @param chat_name name of the chat, for example "Area Chat"
    # @param msg Message to send into the chat
    def msg_to_chat(self, chat_name: str, msg: str):
        self.__session.PBotCharacterAPI().msgToChat(chat_name, msg)

    ## Returns content of each slot in players equipment menu
    # @return list with each slot containing None or PBotItem
    def get_equipment(self) -> List[Optional[PBotItem]]:
        return [PBotItem(x) if x is not None else None for x in self.__session.PBotCharacterAPI().getEquipment()]

    ## Equip item from hand to the given slot
    # @param slot equipment slot to equip item to
    def equipEquipment(self, slot: int):
        self.__session.PBotCharacterAPI().equipEquipment(slot)

File: src/haven/purus/pbot/api/PBotCharacterAPI.java
package haven.purus.pbot.api;

import haven.*;

import java.util.ArrayList;
import java.util.List;

public class
PBotCharacterAPI {

	private PBotSession pBotSession;

	PBotCharacterAPI(PBotSession pBotSession) {
		this.pBotSession = pBotSession;
	}

	/**
	 * Get the player stamina
	 * @return Returns 0-100
	 */
	public double getStamina() {
		return (100 * pBotSession.gui.getmeter("stam", 0).a);
	}

	/**
	 * Get the player energy
	 * @return Returns 0-100
	 */
	public double getEnergy() {
		return (100 * pBotSession.gui.getmeter("nrj", 0).a);
	}

	/**
	 * Get the player hp
	 * @return Returns 0-100
	 */
	public double getHp() {
		return (100 * pBotSession.gui.getmeter("hp", 1).a);
	}

	/**
	 * Send act message to server
	 * Act can be used for example to choose a cursor
	 * Some acts:
	 * dig, mine, carry, destroy, fish, inspect, repair, crime, swim, tracking, aggro, shoot
	 * @param act Act(s) to choose
	 */
	public void doAct(List<String> act) {
		pBotSession.gui.menu.wdgmsg("act", act.toArray());
	}

	/**
	 * Get name of the selected cursor
	 * @return cursor
	 */
	public String getCursName() {
		try {
			return pBotSession.gui.map.ui.getcurs(pBotSession.gui.map.mouseLoc).name;
		} catch(Exception e) {
			return null;
		}
	}

	/**
	 * Cancels the current act by clicking
	 */
	public void cancelAct() {
		pBotSession.gui.map.wdgmsg("click", Coord.z, Coord.z, 3, 0);
	}

	/**
	 * Log out to character selection
	 */
	public void logoutChar() {
		pBotSession.gui.act("lo", "cs");
	}


	/**
	 * Log in with a character from character login menu
	 * @param charname Name of the character
	 */
	public void loginChar(String charname) {
		// TODO
	}

	/**
	 * Set player speed setting
	 * @param speed 1 = crawl, 2 = walk, 3 = run, 4 = sprint
	 */
	public void setSpeed(int speed) {
		Speedget sg = (Speedget) pBotSession.gui.ulpanel.getchild(Speedget.class);
		if(sg != null)
			sg.SpeedToSet = speed;
	}

	/**
	 * Get current speed setting of player
	 * @return 1 = crawl, 2 = walk, 3 = run, 4 = sprint
	 */
	public int getSpeed() {
		Speedget sg = (Speedget) pBotSession.gui.ulpanel.getchild(Speedget.class);
		if(sg != null)
			return sg.cur;
		return -1;
	}

	/**
	 * Get maximum speed setting that player can be set to
	 * @return 1 = crawl, 2 = walk, 3 = run, 4 = sprint
	 */
	public int getMaxSpeed() {
		Speedget sg = (Speedget) pBotSession.gui.ulpanel.getchild(Speedget.class);
		if(sg != null)
			return sg.max;
		return -1;
	}

	/**
	 * Send message to given chat
	 * @param chatName Name of the chat, for example "Area Chat"
	 * @param msg Message to send into the chat
	 */
	public void msgToChat(String chatName, String msg) {
		for(Widget w = pBotSession.gui.chat.lchild; w != null; w = w.prev) {
			if(w instanceof ChatUI.EntryChannel) {
				ChatUI.EntryChannel cht = (ChatUI.EntryChannel) w;
				if(cht.name().equals(chatName))
					cht.send(msg);
			}
		}
	}

	/**
	 * Returns content of each slot in players equipment menu
	 * @return List with each slot containing null or pbotitem
	 */
	public List<PBotItem> getEquipment() {
		Window equwnd = pBotSession.gui.equwnd;
		if(equwnd == null)
			return null;
		Equipory eq = equwnd.getchild(Equipory.class);
		if(eq == null)
			return null;
		ArrayList<PBotItem> ret = new ArrayList<>();
		for(GItem itm : eq.slots) {
			if(itm == null)
				ret.add(null);
			else
				ret.add(new PBotItem(itm, pBotSession));
		}
		return ret;
	}

	/**
	 * Equip item in hand to the given slot
	 * @param slot equipment slot to equip item to
	 */
	public void equipEquipment(int slot) {
		Window equwnd = pBotSession.gui.equwnd;
		if(equwnd != null) {
			Equipory eq = equwnd.getchild(Equipory.class);
			if(eq != null)
				eq.wdgmsg("drop", slot);
		}
	}

}
